core/src/main/java/org/elasticsearch/index/mapper/geo/GeoShapeFieldMapper.java

```diff
@@ -108,91 +112,30 @@
-        }
-
-        public Builder strategy(String strategy) {
-            this.strategyName = strategy;
-            return this;
-        }
-
-        public Builder treeLevelsByDistance(double meters) {
-            this.precisionInMeters = meters;
-            return this;
-        }
-
-        public Builder treeLevels(int treeLevels) {
-            this.treeLevels = treeLevels;
-            return this;
-        }
-
-        public Builder distanceErrorPct(double distanceErrorPct) {
-            this.distanceErrorPct = distanceErrorPct;
-            return this;
-        }
-
-        public Builder orientation(Orientation orientation) {
-            this.orientation = orientation;
-            return this;
+        public GeoShapeFieldType fieldType() {
+            return (GeoShapeFieldType)fieldType;
         }
 
         @Override
         public GeoShapeFieldMapper build(BuilderContext context) {
+            GeoShapeFieldType geoShapeFieldType = (GeoShapeFieldType)fieldType;
 
-            if (Names.TREE_GEOHASH.equals(tree)) {
-                prefixTree = new GeohashPrefixTree(ShapeBuilder.SPATIAL_CONTEXT, getLevels(treeLevels, precisionInMeters, Defaults.GEOHASH_LEVELS, true));
-            } else if (Names.TREE_QUADTREE.equals(tree)) {
-                if (context.indexCreatedVersion().before(Version.V_1_6_0)) {
-                    prefixTree = new QuadPrefixTree(ShapeBuilder.SPATIAL_CONTEXT, getLevels(treeLevels, precisionInMeters, Defaults
-                            .QUADTREE_LEVELS, false));
-                } else {
-                    prefixTree = new PackedQuadPrefixTree(ShapeBuilder.SPATIAL_CONTEXT, getLevels(treeLevels, precisionInMeters, Defaults
-                            .QUADTREE_LEVELS, false));
-                }
-            } else {
-                throw new IllegalArgumentException("Unknown prefix tree type [" + tree + "]");
+            if (geoShapeFieldType.tree.equals("quadtree") && context.indexCreatedVersion().before(Version.V_2_0_0)) {
+                geoShapeFieldType.setTree("legacyquadtree");
             }
-            setupFieldType(context);
-
-            RecursivePrefixTreeStrategy recursiveStrategy = new RecursivePrefixTreeStrategy(prefixTree, fieldType.names().indexName());
-            recursiveStrategy.setDistErrPct(distanceErrorPct);
-            recursiveStrategy.setPruneLeafyBranches(false);
-            TermQueryPrefixTreeStrategy termStrategy = new TermQueryPrefixTreeStrategy(prefixTree, fieldType.names().indexName());
-            termStrategy.setDistErrPct(distanceErrorPct);
 
-            GeoShapeFieldType geoShapeFieldType = (GeoShapeFieldType)fieldType;
-            geoShapeFieldType.setStrategies(strategyName, recursiveStrategy, termStrategy);
-            geoShapeFieldType.setOrientation(orientation);
+            if (context.indexCreatedVersion().before(Version.V_2_0_0) ||
+                (geoShapeFieldType.treeLevels() == 0 && geoShapeFieldType.precisionInMeters() < 0)) {
```

Comment:
I know this was here before, definitely not a blocker as we can clean it up in another PR, but this treeLevels == 0 check has bothered me. Especially since a treeLevel of 0 doesn't make sense and this is only used as an "initial state" condition. It seems cleaner to set the default tree and level at construction and appropriately change if/when the user explicitly defines the tree?  That might make these conditionals cleaner and the whole tree parameter code easier to read?